agent/structs/discovery_chain.go

```diff
@@ -183,17 +181,21 @@ var _ encoding.TextUnmarshaler = (*DiscoveryTarget)(nil)
 //
 // This should NOT return any errors.
 func (t DiscoveryTarget) MarshalText() (text []byte, err error) {
+	return []byte(t.Identifier()), nil
+}
+
+func (t DiscoveryTarget) Identifier() string {
 	var buf bytes.Buffer
 	buf.WriteString(url.QueryEscape(t.Service))
 	buf.WriteRune(',')
-	buf.WriteString(url.QueryEscape(t.ServiceSubset))
+	buf.WriteString(url.QueryEscape(t.ServiceSubset)) // TODO(rb): move this first so the scoping flows from small->large?
```

Comment:
If this is empty we end up with something like `web,,dc1` or `web,,namespace,dc1`. Is there a reason to make namespace be optional but not subset? I'd expect subset empty to be 99.9% of cases.

Is there an argument to format this as `subset.name.namespace.dc` to match SNI or reverse that to match natural precedence `dc/namespace/name/subset` or something?

All bikesheds as none really affect the behaviour just seems we should do one that makes sense. Is there a specific reason it's URI encoded?

Reply:
The query encoding was because I wasn't entirely sure what was and was not a reserved character for 3 of the 4 components (I know I validated subsets).

This encoding predates the bulk envoy naming SNI changes. And given _no_ escaping happens there already then there's no point of escaping stuff here. I actually was thinking of switching the DiscoveryTarget marshalling over to the SNI variant already.

Reply:
Going to handle this in https://github.com/hashicorp/consul/pull/6248 instead.